File: src/cpp/HybridSimAgent.cpp
#include "HybridSimAgent.h"

void HybridSimAgent::calculate(bool is_opponent) {
	if (currentNode->childStates == nullptr) {
		expand();
	}
	else {
		nodeSelection(is_opponent);
	}
}

void HybridSimAgent::nodeSelection(bool opponent) {
	if (opponent) {
		double sum = 0;
		for (const StateNode* s : *currentNode->childStates) {
			sum += (*s).enemyValue;
		}
		const double c_value = sum * uniform_0_1(srs);
		sum = 0;
		StateNode* selectedState = nullptr;
		for (StateNode* s : *currentNode->childStates) {
			sum += (*s).enemyValue;
			if (sum >= c_value) {
				selectedState = s;
				break;
			}
		}
		nextNode = selectedState;
		return;
	}

	double sum = 0;
	for (const StateNode* s : *currentNode->childStates) {
		sum += (*s).selfValue;
	}
	const double c_value = sum * uniform_0_1(srs);
	sum = 0;
	StateNode* selectedState = nullptr;
	for (StateNode* s : *currentNode->childStates) {
		sum += (*s).selfValue;
		if (sum >= c_value) {
			selectedState = s;
			break;
		}
	}
	nextNode = selectedState;
}

void HybridSimAgent::expand() {
	MatrixArray nextNodes;
	const int crd = currentNode->depth;
	if (crd < 10)
	{
		int rev_size = 1500 - crd * 150;
		rev_size = max(rev_size, 550);
		nextNodes.reserve(rev_size);
	}
	else if (crd < 15)
	{
		nextNodes.reserve(250);
	}
	getPossibleStates(nextNodes, currentNode->mat, side);

	const u64 size_i = nextNodes.size();
	OrderedPairList valueList(size_i);
	for (int i = 0; i < size_i; i++) {
		valueList[i].x = i;
		valueList[i].y = 0;
		valueList[i].y += numPossibleMoves(nextNodes[i], side);
		valueList[i].y -= numPossibleMoves(nextNodes[i], enemy_side);
	}
	OrderedPairList::sort_by_y(valueList);
	OrderedPairList betterList;
	OrderedPairList::front(betterList, valueList, 32);
	if (betterList.empty()) {
		nextNode = currentNode;
		return;
	}

	MatrixArray top_mat;
	const u64 size_j = betterList.size();
	for (int i = 0; i < size_j; i++) {
		top_mat.push_back(nextNodes[betterList[i].x]);
	}

	auto* top_states = new StateNodesArray(top_mat, *mem_pool);
	sna_list.push_back(top_states);
	currentNode->childStates = top_states;

	for (StateNode* state : *(currentNode->childStates)) {
		state->parentState = currentNode;
		state->depth++;
		if (turn_num < 5)
			state->gamma = 0.80 - 0.01 * turn_num;
		else if (turn_num < 10)
			state->gamma = 0.70;
		else if (turn_num < 30)
			state->gamma = 0.66;
		else
			state->gamma = 0.63;
	}
	const int x = (int)((*top_states).size() * uniform_0_1(srs));
	nextNode = (*top_states).at(x);
}

void HybridSimAgent::update(StateNode* s) {
	currentNode = s;
}

OrderedPairList HybridSimAgent::getAction() const {
	return parseAction(currentNode->mat, nextNode->mat);
}

File: src/src/amazon_ai/hybrid_mcts/HybridSimAgent.java
package amazon_ai.hybrid_mcts;

import amazon_framework.AbstractAIAgent;
import amazon_framework.MatrixArray;
import amazon_framework.OrderedPair;
import amazon_framework.OrderedPairList;

import java.util.Random;

import static amazon_framework.Operations.*;

public final class HybridSimAgent extends AbstractAIAgent {
    StateNode currentNode;
    StateNode nextNode;
    int num;

    HybridSimAgent(StateNode inputState, byte side) {
        super(inputState.mat, side);
        currentNode = inputState;
        currentNode.mat = inputState.mat;
    }

    public void calculate(boolean is_opponent) {
        if (currentNode.childStates == null) {
            expand();
        } else {
            nodeSelection(is_opponent);
        }
    }

    public void nodeSelection(boolean opponent) {
        if (!opponent) {
            double sum = 0;
            for (StateNode s : currentNode.childStates) {
                sum += s.selfValue;
            }
            double cvalue = sum * Math.random();
            StateNode selectedState = null;
            sum = 0;
            for (StateNode s : currentNode.childStates) {
                sum += s.selfValue;
                if (sum >= cvalue) {
                    selectedState = s;
                    break;
                }
            }
            nextNode = selectedState;
        }

        if (opponent) {
            double sum = 0;
            for (StateNode s : currentNode.childStates) {
                sum += s.enemyValue;
            }
            double cvalue = sum * Math.random();
            StateNode selectedState = null;
            sum = 0;
            for (StateNode s : currentNode.childStates) {
                sum += s.enemyValue;
                if (sum >= cvalue) {
                    selectedState = s;
                    break;
                }
            }
            nextNode = selectedState;
        }
    }

    public void expand() {
        MatrixArray nextNodes = getPossibleStates(currentNode.mat, side);
        OrderedPairList valueList = new OrderedPairList();
        final int size_i = nextNodes.size();
        for (int i = 0; i < size_i; i++) {
            OrderedPair value = new OrderedPair();
            value.x = i;
            value.y += numPossibleMoves(nextNodes.at(i), side);
            value.y -= numPossibleMoves(nextNodes.at(i), enemy_side);
            valueList.add(value);
        }
        valueList.sort_by_y();
        OrderedPairList betterList = valueList.back(32);

        if (betterList.isEmpty()) {
            nextNode = currentNode;
            return;
        }

        MatrixArray marr = new MatrixArray();

        for (int i = 0; i < betterList.size(); i++) {
            int index = betterList.at(i).x;
            marr.add(matrixCopy(nextNodes.at(index)));
        }

        StateNodesArray sarr = new StateNodesArray(marr);

        currentNode.childStates = sarr;

        for (StateNode s : currentNode.childStates) {
            s.parentState = currentNode;
        }

        Random rand = new Random();
        final int x = rand.nextInt(sarr.size());
        nextNode = sarr.at(x);
    }

    public void update(StateNode s) {
        currentNode = s;
    }

    @Override
    public OrderedPairList getAction() {
        return parseAction(currentNode.mat, nextNode.mat);
    }
}